### PR TITLE
Initial draft

### PR DESCRIPTION
**Pull request by zcorpan**
A start of milestone 0 in https://github.com/w3c/aria-at-automation/issues/15

This is modeled after the WebDriver BiDi spec: https://w3c.github.io/webdriver-bidi/

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/aria-at-automation/pull/19.html" title="Last updated on Aug 16, 2022, 12:23 PM UTC (670fcab)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/aria-at-automation/19/22d3c12...670fcab.html" title="Last updated on Aug 16, 2022, 12:23 PM UTC (670fcab)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -4,14 +4,88 @@ Shortname: aria-at-automation
...
+AT Driver defines a protocol for introspection and remote control of assistive technology software, using a bidirectional communication channel.
+
+Explainer {#explainer}
+======================
+
+Specify a protocol using WebSocket that maximally reuses concepts and conventions from WebDriver BiDi.
+
+A connection has two endpoints: remote and local. The remote end can control and read from the screen reader, which can either be implemented as a standalone application or be implemented as part of the AT software. The local end is what the test interfaces with, usually in the form of language-specific libraries providing an API.
+
+(For WebDriver, you still need to use HTTP to request the "webSocketUrl" capability, and then establish the WebSocket connection. Here we want the WebSocket connection to be the initial and only connection.)
```
Comment (jugglinmike) [inline index.bs:26]: We should take care with wording here since WebSocket necessarily begins with an HTTP handshake.

**Comment by jscholes** [comment]
@zcorpan Some questions that occured to me; these don't necessarily need answers right now but would be good to note down as TODOs:

1. Does, or should, the protocol imply anything about the ordering of data coming from the remote end?  E.g. if I send a command, and am awaiting a response, can an event-related payload be sent back before that response?
2. Will the transport use SSL?  Is this something the protocol itself should concern itself with?

**Comment by zcorpan** [comment]
> @zcorpan Some questions that occured to me; these don't necessarily need answers right now but would be good to note down as TODOs:
> 
> 1. Does, or should, the protocol imply anything about the ordering of data coming from the remote end? E.g. if I send a command, and am awaiting a response, can an event-related payload be sent back before that response?

I think this should be the same as in WebDriver BiDi:

> Multiple commands can run at the same time, and commands can potentially be long-running. As a consequence, commands can finish out-of-order.

https://w3c.github.io/webdriver-bidi/#commands

There is no restriction in when a remote end can send events, and so they could be sent in between a command message and its result message.

> 2. Will the transport use SSL?  Is this something the protocol itself should concern itself with?

WebSocket can be used with or without TLS. WebDriver BiDi also allows both to be used: https://w3c.github.io/webdriver-bidi/#listener-secure-flag

At least for connections that are over the wire, it seems good to at least strongly suggest TLS connections to be used, for both privacy and security. WebDriver BiDi doesn't currently say anything about this as far as I can tell.

**Comment by zcorpan** [comment]
Now the draft contains the first command `session.new` where you can request a session of a particular AT with a specific version on a specific platform, or a choice of several (`firstMatch`). The result will contain the chosen AT's name, version, and platform.

**Comment by zcorpan** [comment]
> If each AT will host its own remote end, acting as the server, the client end will need to already decide which AT to connect to before establishing the transport.

It's possible to implement a remote end that supports multiple ATs, or multiple versions of an AT.

WebDriver also supports proxies, which could be applicable for us as well but isn't included in the draft currently.

That's not to say that all implementations have to support more than one version of one AT on one platform. But the protocol can be designed to allow scaling.

That said, we can remove `firstMatch` for now, to keep it simple.

>  If I've already decided to connect to NVDA, asking for NVDA on Windows seems redundant because I'll already have it, and won't have access to anything else.

Yes, it is redundant in that case. Though I think it seems worthwhile to have it so that it's consistent with WebDriver and we can support remote ends controlling multiple versions or multiple ATs going forward.

> Additional question: platformName is defined as text, but should it have any specific format/level of detail? E.g. "Windows", "Windows 10", "windows" in lower case, something including a build number/identifier, etc.?

In WebDriver, it's lowercase and some suggested values are "windows", "macos", "linux".

https://w3c.github.io/webdriver/#dfn-matching-capabilities

There used to be a `platformVersion` as well but it was removed from the WebDriver spec, see https://github.com/w3c/webdriver/issues/741 .

File: index.bs [hunk 2]
```diff
@@ -2,16 +2,255 @@
...
+
+Each [=command=] is defined by:
+
+* A <dfn>command type</dfn> which is defined by a [=remote end definition=] fragment containing a group. Each such group has two fields:
+  * `method` which is a string literal in the form `[module name].[method name]`. This is the <dfn>command name</dfn>.
+  * `params` which defines a mapping containing data that to be passed into the command. The populated value of this map is the <dfn>command parameters</dfn>.
+* A <dfn>result type</dfn>, which is defined by the [=local end definition=] fragment.
+* A set of [=remote end steps=] which define the actions to take for a command given a [=session=] and [=command parameters=] and return an instance of the command [=return type=].
+
+A command that cna run without an active session is a <dfn>static command</dfn>. Commands are not static commands unless stated in their definition.
```
Comment (mfairchild365) [inline index.bs:117]: I think "cna" here should be "can"

File: index.bs [hunk 3]
```diff
@@ -2,16 +2,255 @@
...
+
+When commands are sent from the [=local end=] they have a command id. This is an identifier used by the [=local end=] to identify the response from a particular command. From the point of view of the [=remote end=] this identifier is opaque and cannot be used internally to identify the command.
+
+Events {#protocol-events}
+-------------------------
+
+An <dfn>event</dfn> is a notification, sent by the [=remote end=] to the [=local end=], signaling that something of interest has occurred oin the [=remote end=].
```
Comment (mfairchild365) [inline index.bs:124]: I think "oin" should be "in"

File: index.bs [hunk 4]
```diff
@@ -2,16 +2,255 @@
...
+
+* An <dfn>event type</dfn> is defined by a [=local end definition=] fragment containing a group. Each such group has two fields:
+  * `method` which is a string literal of the form `[module name].[event name]`. This is the <dfn>event name</dfn>.
+  * `params` which defins a mapping containing event data. The populated value of this map is the <dfn>event parameters</dfn>.
```
Comment (mfairchild365) [inline index.bs:128]: I think "defins" should be "defines"

**Comment by zcorpan** [comment]
I've committed some more progress here today. The `session.new` command is not fully specced out, but there's a lot more now than previously, so may be worth a review at this point.

**Comment by zcorpan** [comment]
In https://github.com/w3c/aria-at-automation/pull/19/commits/57c161d3e2505a371d65d0f1183fcef341691587 I added some code to extract the CDDL into two files (`at-driver-remote.cddl` and `at-driver-local.cddl`), using `npm run extract-schemas`.

No JSON Schemas yet.

**Comment by zcorpan** [comment]
@SamuelHShaw @jscholes you said in our check-in meeting yesterday that PAC have reviewed this and had no comments. Can you mark it as reviewed with GitHub's UI? Then we can merge. Thanks!

**Comment by jugglinmike** [comment]
Thanks to feedback from @jkva in gh-26, I've fixed a few typos and removed the `assert` implementation. (If the need arises, it might be better to rely on [Node.js's built-in `assert` module](https://nodejs.org/api/assert.html) for that functionality.)

**Pull request merged**